lib/MonadicRewrite.thy

```diff
@@ -524,13 +524,19 @@ lemma monadic_rewrite_assert2:
    \<Longrightarrow> monadic_rewrite F E ((\<lambda>s. Q \<longrightarrow> P s) and (\<lambda>_. Q)) (assert Q >>= f) g"
   by (auto simp add: assert_def monadic_rewrite_def fail_def split: if_split)
 
-lemma monadic_rewrite_state_assert_true:
-  "monadic_rewrite F E P (state_assert P) (return ())"
-  by (simp add: state_assert_def monadic_rewrite_def exec_get)
+(* under assumption of non-failure of LHS, we can assume any assertions for showing preconditions
+   of rewrite *)
+lemma monadic_rewrite_state_assert:
+  "monadic_rewrite F E P (f ()) g
+   \<Longrightarrow> monadic_rewrite F E (\<lambda>s. (F \<longrightarrow> Q s \<longrightarrow> P s) \<and> (\<not>F \<longrightarrow> P s \<and> Q s))
+                      (state_assert Q >>= f) g"
```

Comment:
Would it add any convenience later in the proofs if we provided two versions where `F` is instantiated (and the guard therefore simplifies a bit), or does this disappear anyway before the user really sees it?

Reply:
I thought about it, but didn't see any point. If you apply this rule, you end up with a wp goal later on, which will be cleared with a `simp` given any specific value of `F`. If an extreme situation, they can do `where F=..., simplified`, but again, not seeing it.